feat: E-commerce Refactor

marination opened the pull request:
- [x]  Create **Website Item** to handle all Website Related Data
- [x] Publish Item from Item Master (create Website Item)
- [x] Consolidated **E-commerce Settings** (replace Products and Shopping Cart Settings)
- [x] Drive Portal via **Website Item** and **E Commerce Settings** only (cleanup item and make stuff obsolete)
- [x] Patch for:  Items ➡️   Website Items  (needs load testing)
- [x] Patch: Shopping Cart & Products Settings ➡️    E Commerce Settings
- [x] Patch: Homepage Products :- Items ➡️    Website Items 
- [x] Fix Tests

------
- [x] All Categories page (configurable)
- [x] Wishlist/Add to Cart (Card Actions) 
- [x] Wishlist Full Page
- [x] Item Full Page (better arrangement of info, add to wishlist) 
- [x] Ratings and Reviews
- [x] Capturing discounts
- [x] Shopping Cart Page (remove items in a click and bug fixes)
- [x] Grid and List view Toggling (Listing Page)
- [x] Search
- [x] Server side Test cases
- [x] Review changes
-----------------------------------

Redisearch Docs: https://github.com/frappe/erpnext_documentation/pull/375
[Official Documentation](https://docs.erpnext.com/docs/v13/user/manual/en/e_commerce/set_up_e_commerce)

[Unofficial E-commerce guide](https://docs.google.com/document/d/17s6VxWeihTImSG2toR5Ma1jNq9nNAaYueP-dimNdQ4I/edit?usp=sharing)

marination committed:
feat: E-commerce Refactor

- Created "Website Item" to handle website related information
- Publishing Item new flow
- Created "E Commerce Settings"
- Removed Products Settings

marination committed:
chore: Drive E-commerce via Website Item

- Removed Shopping Cart Settings
- Portal fully driven via E Commerce Settings
- All Item listing querying will happen via ProductQuery engine only
- Product Listing via Website Items
- removed redundant code
- Moved all website logic from Item to Website Item

marination committed:
chore: Removing Item's Website section & references

- Removed old onboarding slide json and `create_onboarding_docs` methods
- Removed website sections from Item master
- Removed references to item website fields
- Shifted Item doctype website methods to Website Item
- Removed WebsiteGenerator from Item doctype
- Website Items in Homepage Products section
- Removed redundant code from item_group.py
- Fix: Item field filters won’t appear in website

marination committed:
fix: Sider and indexing (minor)

marination committed:
fix: Hide Attribute filters if 'Hide Variants' is enabled in E Commerce Settings

- Hide Attribute filters if 'Hide Variants' is enabled in E Commerce Settings
- Consider 'Hide Variants' in ProductQuery Engine
- Added docstrings
- Remove `get_e_commerce_settings`, redundant

marination committed:
chore: Patches for Website Item

- Patch to make website item from item
- Patch to move Products and Shopping Cart Settings into E Commerce Settings
- Patch to move products in Homepage from Item to Website Item
- Minor cleanup, replacing/removing references to Item website fields (obsolete)

mujeerhashmi commented:
Recommendation:
Qty, Item Attributes (Size, Colour) selection should be on the product page itself above `Add to Cart` unlike the present pop-up for item attribute selection.

marination committed:
chore: Removed Shopping Cart Module

- Moved all files and web templates from Shopping Cart to E-commerce module
- Made Shopping Cart module obsolete
- Moved select E-commerce related files from Portal to E-commerce module
- Minor cleanups
- Fixed Shopping Cart and Product Configurator tests

marination commented:
![Screenshot 2021-02-25 at 1 51 28 PM](https://user-images.githubusercontent.com/25857446/109125697-5e7a9100-7772-11eb-9697-06a572937c19.png)

marination committed:
feat: Shop by Category

- Added Shop by Category Page
- Tabbed sections for item fields in Shop by Category Page
- Added Shop by Category Section in E commerce Settings
- Nested Navigation & Breadcrumbs in Item group pages
- Added scrollable & clickable Sub categories in Item Group page
- Made breadcrumbs slightly dynamic in Item Page
- Added image to Brand doctype

marination committed:
feat: Card Actions and Wishlist

- Rough UI for card actions
- Wishlist doctype
- Indicators on card based on stock availability

marination committed:
feat: Animate Add to Cart List interactions (UX)

- Increased qty in cart on clicking add to cart for existing item
- Simplified macro arguments
- Navbar cart icon animation
- Explore button for template item in card
- Add to cart button animation

marination committed:
Merge branch 'develop' into e-commerce-refactor

marination committed:
feat: Wishlist from card actions

- Add remove items from wishlist
- Wishlist icon at nav bar
- Animate wishlist icon in card and navbar
- Remember wished state after refresh as well

marination committed:
feat: Wishlist Page

- Navbar icon with badge count for wishlist
- Wishlist page with cards
- Cards can be moved to cart or removed in a click
- Separated all wishlist related methods into wishlist.js
- Made a common js method(util) to add/remove wishlist items
- Bug fix: Make sure items are removed from session user's wishlist

marination committed:
fix: Sider

marination committed:
Merge branch 'develop' into e-commerce-refactor

marination committed:
feat: Product Details Tabbed Section and Add to Wishlist in Item Full Page

- Add to Wishlist button next to add to cart
- Beautified Product Specifications table section
- Product Specifications can be optionally in a tabbed section with custom tabs added
- Removed hard coded gray bg to allow custom theme overwrites
- Fixed resizing issues in Item Full Page view
- Cleaned up inline styles and ported to scss

marination committed:
fix: Remove unnecessary css variable and hover state

- Removed wish-red variable
- Removed hover state from remove wishlist item button in card
- Removed inline css from wishlist item card

mujeerhashmi commented:
We also need a provision to define the actual rate and then discounted rate. The actual rate should be struck off and discount % next to it should be shown.

rtdany10 commented:
> We also need a provision to define the actual rate and then discounted rate. The actual rate should be struck off and discount % next to it should be shown.

Agreed.
Good to see the shopping cart reworked!🥳

rtdany10 commented:
> Recommendation:
> Qty, Item Attributes (Size, Colour) selection should be on the product page itself above `Add to Cart` unlike the present pop-up for item attribute selection.

Agreed.
Good to see the cart reworked🥳

marination committed:
feat: (wip) Ratings and Reviews

- Added Ratings and Reviews section in Item full page view
- Added provision to write a review with popup
- Created Item Review Doctype to store User-Item unique reviews
- Added privision to enable/disable wishlist and reviews in e commerce settings
- Hide cart and wishlist actions everywhere (even navbar) depending on settings
- Moved some more inline css to scss
- Small logic fixes
TODO: Reviews full page view with paging

barredterra commented:
Last time i checked ecommerce supported only one payment method. Would it be within the scope of this. PR to allow multiple payment methods?

marination committed:
feat: Customer Ratings & Reviews Full Page

- Created macros for repetitive snippets
- Created Customer Reviews full page
- View more button to reveal 10 more reviews at a time
- Common function to get reviews with start and end

marination commented:
@mujeerhashmi variant config ux and slashed prices are being worked upon
@barredterra sadly not in this PR, but will give it thought after this is done

marination committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into e-commerce-refactor

marination committed:
fix: Sider and Tests

Alchez commented:
@marination @barredterra 

We had attempted to allow multiple payment gateways before with #21374, but that went nowhere. Maybe something can be salvaged from there? :sweat_smile:

marination committed:
feat: Slashed Prices and Discount display

- Registered mrp and price after discounts
- slashed price with discount in listing, item page and wishlist
- removed redundant imports
- renamed method to `get_web_item_qty_in_stock` to get Website Item stock
- adjusted styles for resizing
- made add to cart button full width on cards

marination committed:
fix: Item schema modified date update

- Modified date changed due to merge conflict
- Changed it to a later date to make sure Item migrates correctly

marination committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into e-commerce-refactor

marination committed:
feat: Discount Filters

- Discount filters in filters section
- Code cleanup

rtdany10 commented:
Would be nice to set a default Return & Refund Policy and also set different ones for individual items(if needed) and display them under each item.

marination commented:
@rtdany10 there's already provision for that, will update screenshots soon perhaps

marination committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into e-commerce-refactor

marination committed:
chore: Sider and Semgrep

NagariaHussain committed:
feat: Add basic autocomplete using redisearch

NagariaHussain committed:
feat: Basic Query + Autocomplete

NagariaHussain committed:
feat: Add search fields field

NagariaHussain committed:
feat: Make search index fields configurable

- Move indexing logic to separate file
- Add more validation logic for 'search index fields' field

NagariaHussain committed:
feat: Add Category autocomplete with config in settings

NagariaHussain committed:
chore: Make it a little beautiful

NagariaHussain committed:
refactor: Use global redis connection

NagariaHussain committed:
chore: Add query clean-up

NagariaHussain committed:
chore: Add placeholder image

NagariaHussain committed:
feat: Add brand line display setting

NagariaHussain committed:
feat: Show brand line in search results

NagariaHussain committed:
feat: Show Recent Searches

marination committed:
feat: Offer Display

- Added offers section in website item
- Added more roles to website item
- Fixed attachment limit issue in website item
- Created Website Offer child doctype
- Added offers listing in Item full page view
- style fixes

marination committed:
fix: Show Offers section only if offers exist

NagariaHussain committed:
chores: Add function params and remove unused imports

marination committed:
feat: (wip) Toggle Views

- Auto Height on Cards
- Title with ellipses on length exceed
- Changed namepaces
- Moved product card rendering to JS
- Added Image and List View Toggling buttons
- Kept basic filters rendering just as before

marination committed:
Merge branch 'version-13-hotfix' into e-commerce-refactor

marination committed:
feat: Product View toggling

- Added fully functional list and grid view toggling
- Added ProductGrid and ProductList controllers
- Moved html snippets, rendered via JS now
- Item Group page also rendered via common controller
- Paging section rendered via JS
- Minor style changes

marination committed:
fix: Sider and Patches

- Sider fixes
- Deleted patches that worked as per old schema
- Cleared instances of web fields linked to Item master

rtdany10 commented:
In the product page, it shows stock correctly, but the user can increase the quantity to any number after adding it to the cart. Even if the stock is 3, the user can get quotation for up to 5 or 6 or n numbers. It would be nice to have a check while increasing the quantity from the cart page.

(not sure if it's fixed in this one. experienced this issue in the current cart)

marination committed:
fix: Patch fix for fresh installs

- minor sider fix
- patch fix for fresh installs

marination committed:
fix: Empty states and miscellanous fixes

- Added Wishlist and Product Listing empty states
- Hide ‘Write Review’ button f user is not Customer
- Fixed grid view title (missing arg)
- Render empty state if no items form server side
- Removed unused function
- Guest user redirected to login on clicking Wishlist button
- Fixed ‘Notes’ field clearing issue

marination committed:
fix: Move get_product_filtrs_data to website item file

marination committed:
fix: Items Query, Sub categories, Review Button

- Include Item Variant table in query only if attributes are involved
- Render sub categories of they exist, even without items
- 'Write a Review' provision only if user is a customer

ankush commented on erpnext/e_commerce/product_query.py:
```diff
@@ -0,0 +1,192 @@
+
+	def query_items(self, conditions, or_conditions, substitutions, start=0, with_attributes=False):
+		"""Build a query to fetch Website Items based on field filters."""
+		self.query_fields = (", ").join(self.fields)
```
```suggestion
		self.query_fields = ", ".join(self.fields)
```

ankush commented on erpnext/e_commerce/product_configurator/utils.py:
```diff
@@ -0,0 +1,195 @@
+			next_attribute = attribute
+			break
+
+	valid_options_for_attributes = frappe._dict({})
```
```suggestion
	valid_options_for_attributes = frappe._dict()
```

ankush commented on erpnext/e_commerce/product_configurator/utils.py:
```diff
@@ -0,0 +1,195 @@
+				NULL
+		'''.format(attribute_query=attribute_query, variant_of_query=variant_of_query)
+
+		item_codes = set([r[0] for r in frappe.db.sql(query, query_values)])
```
select `distinct` instead of using `set`?

ankush commented on erpnext/www/shop-by-category/index.py:
```diff
@@ -0,0 +1,73 @@
+
+			categorical_data[category] = frappe.db.sql("""
+				Select {fields}
+				from `tab{doctype}`""".format(doctype=doctype, fields=",".join(fields)), as_dict=1)
```
If formatting is must then f-strings are more readable IMO.
```suggestion
			categorical_data[category] = frappe.db.sql(f"""
				select {",".join(fields)}
				from `tab{doctype}`""", as_dict=1)
```

ankush commented on erpnext/www/shop-by-category/index.py:
```diff
@@ -0,0 +1,73 @@
+		'rounded': 1,
+		'slider_name': "Categories"
+	}
+	slideshow = frappe.get_doc("Website Slideshow", slideshow)
```
```suggestion
	slideshow = frappe.get_cached_doc("Website Slideshow", slideshow)
```

ankush commented on erpnext/www/shop-by-category/index.py:
```diff
@@ -0,0 +1,73 @@
+	}
+	slideshow = frappe.get_doc("Website Slideshow", slideshow)
+	slides = slideshow.get({"doctype": "Website Slideshow Item"})
+	for index, slide in enumerate(slides):
+		values[f"slide_{index + 1}_image"] = slide.image
+		values[f"slide_{index + 1}_title"] = slide.heading
+		values[f"slide_{index + 1}_subtitle"] = slide.description
+		values[f"slide_{index + 1}_theme"] = slide.get("theme") or "Light"
+		values[f"slide_{index + 1}_content_align"] = slide.get("content_align") or "Centre"
+		values[f"slide_{index + 1}_primary_action"] = slide.url
```
```suggestion
	for index, slide in enumerate(slides, start=1):
		values[f"slide_{index}_image"] = slide.image
		values[f"slide_{index}_title"] = slide.heading
		values[f"slide_{index}_subtitle"] = slide.description
		values[f"slide_{index}_theme"] = slide.get("theme") or "Light"
		values[f"slide_{index}_content_align"] = slide.get("content_align") or "Centre"
		values[f"slide_{index}_primary_action"] = slide.url
```

ankush commented on erpnext/www/shop-by-category/index.py:
```diff
@@ -0,0 +1,73 @@
+	}
+
+	categorical_data = get_category_records(categories)
+	for index, tab in enumerate(categorical_data):
+		tab_values[f"tab_{index + 1}_title"] = frappe.unscrub(tab)
+		# pre-render cards for each tab
+		tab_values[f"tab_{index + 1}_content"] = frappe.render_template(
```
```suggestion
	for index, tab in enumerate(categorical_data, start=1):
		tab_values[f"tab_{index}_title"] = frappe.unscrub(tab)
		# pre-render cards for each tab
		tab_values[f"tab_{index}_content"] = frappe.render_template(
```

ankush commented on erpnext/www/shop-by-category/index.py:
```diff
@@ -0,0 +1,73 @@
+sitemap = 1
+
+def get_context(context):
+	settings = frappe.get_doc("E Commerce Settings")
```
```suggestion
	settings = frappe.get_cached_doc("E Commerce Settings")
```

ankush commented on erpnext/templates/pages/wishlist.py:
```diff
@@ -0,0 +1,57 @@
+			from
+				`tabWishlist Items`
+			where
+				parent=%(user)s""" % {"user": frappe.db.escape(frappe.session.user)}, as_dict=1)
```
```suggestion
				parent=%(user)s""", {"user": frappe.session.user}, as_dict=1)
```

ankush commented on erpnext/templates/pages/wishlist.py:
```diff
@@ -0,0 +1,57 @@
+			},
+			"actual_qty")
+	)
+	return True if stock_qty else False
```
```suggestion
	return bool(stock_qty)
```

ankush commented on erpnext/templates/pages/customer_reviews.py:
```diff
@@ -0,0 +1,16 @@
+# License: GNU General Public License v3. See license.txt
+from __future__ import unicode_literals
+
+no_cache = 1
```
Why disable caching here?

ankush commented on erpnext/templates/pages/customer_reviews.py:
```diff
@@ -0,0 +1,16 @@
+# Copyright (c) 2015, Frappe Technologies Pvt. Ltd. and Contributors
+# License: GNU General Public License v3. See license.txt
+from __future__ import unicode_literals
```
```suggestion

```

ankush commented on erpnext/templates/pages/wishlist.py:
```diff
@@ -0,0 +1,57 @@
+# Copyright (c) 2021, Frappe Technologies Pvt. Ltd. and Contributors
+# License: GNU General Public License v3. See license.txt
+from __future__ import unicode_literals
```
```suggestion
```

ankush commented on erpnext/e_commerce/product_query.py:
```diff
@@ -0,0 +1,192 @@
+
+		# Join the meta fields and default fields set
+		search_fields = default_fields.union(meta_fields)
+		try:
+			if frappe.db.count('Item', cache=True) > 50000:
+				search_fields.remove('description')
+		except KeyError:
+			pass
```
```suggestion
		if frappe.db.count('Item', cache=True) > 50000:
			search_fields.discard('description')
```

ankush commented on erpnext/e_commerce/product_query.py:
```diff
@@ -0,0 +1,192 @@
+
+			all_items.append(set(items_dict.keys()))
+
+		result = [items_dict.get(item) for item in list(set.intersection(*all_items))]
```
```suggestion
		result = [items_dict.get(item) for item in set.intersection(*all_items)]
```

ankush reviewed (commented):
some minor nitpicks

marination committed:
fix: Discount Filters and Web templates

- Fixed discount filters (didn’t work after js render change)
- Fix Item Card Group template height and style
- Add placeholder to missing images in Product Category Cards template
- Code cleanup

NagariaHussain committed:
feat: Add decorator for redisearch

NagariaHussain committed:
chore: Redisearch module load tracking

marination committed:
chore: Shopping Cart styles and cleanup shallow

- Added remove button to cart item rows
- Freeze on change in Shopping Cart (UX)
- Fixed cart items and taxes/totals alignment issues
- Made Cart responsive
- Added free item indicator
- Fixed item group nested routing issue
- Sales Order is populated with right source warehouse

marination committed:
Merge branch 'version-13-hotfix' into e-commerce-refactor

NagariaHussain committed:
feat: Add fallback when redisearch not loaded

NagariaHussain committed:
chore: Redisearch warning in search settings

NagariaHussain committed:
fix: Documentation link and open in new tab

NagariaHussain committed:
Merge branch 'search-with-redisearch' into e-commerce-refactor

ankush commented on erpnext/e_commerce/website_item_indexing.py:
```diff
@@ -0,0 +1,214 @@
+    out = cache.execute_command('MODULE LIST')
+
+    parsed_output = " ".join(
+        (" ".join([s.decode() for s in o if not isinstance(s, int)]) for o in out)
```
```suggestion
        (" ".join(s.decode() for s in o if not isinstance(s, int)) for o in out)
```

ankush commented on erpnext/e_commerce/website_item_indexing.py:
```diff
@@ -0,0 +1,214 @@
+
+    return "search" in parsed_output
+
+# Decorator for checking wether Redisearch is there or not
```
Put this in docstring.

ankush commented on erpnext/e_commerce/website_item_indexing.py:
```diff
@@ -0,0 +1,214 @@
+    return "search" in parsed_output
+
+# Decorator for checking wether Redisearch is there or not
+def redisearch_decorator(function):
```
```suggestion
def redisearch(function):
```

The function name should convey the use case. No need to put "decorator" in name.  Comment above the declaration should be docstring instead.

gavindsouza replied:
The name should convey the use case +1. That's why just  `redisearch` or `redisearch_decorator` are bad names. Perhaps something along the lines of `is_redisearch_loaded`?

**Edit:** Or...how about `if_redisearch_loaded` 😳

ankush commented on erpnext/e_commerce/website_item_indexing.py:
```diff
@@ -0,0 +1,214 @@
+def create_web_item_map(website_item_doc):
+	fields_to_index = get_fields_indexed()
+
+	web_item = {}
+
+	for f in fields_to_index:
+		web_item[f] = website_item_doc.get(f) or ''
```
This can be converted to single comprehension.

ankush commented on erpnext/e_commerce/doctype/e_commerce_settings/e_commerce_settings.py:
```diff
@@ -51,6 +56,33 @@ def validate_checkout(self):
+			return 
+		
+		# Clean up
+		# Remove whitespaces
```
Remove comments where the code is self-explanatory.

marination committed:
feat: Search UI

- Search UI with dropdown results
- Client class to handle Product Search actions and results
- Integrated Search bar into all-products and item group pages
- Run db search without redisearch
- Cleanup: [Search] change decorator names and variables
- Sider fixes

marination committed:
Merge branch 'version-13-hotfix' of https://github.com/frappe/erpnext into e-commerce-refactor

marination committed:
fix: Sider and Patch tests

marination committed:
fix: Search UI open/close and Empty product search results

rtdany10 commented:
Recently, we applied for PayTM business account and they went through our cart. They pointed out that the customer had the option to choose Country while creating their address and hence we were asked to show our Import/Export license. While in reality, we don't even ship outside state.

Would be nice if the new cart has an option which lets the system manager control which all data to collect for customer address.
(A work around would be to keep the default value of country to India and make it read-only, but can't do that.)

marination committed:
fix: Mixin import issue, search error handling, shop-by-category slideshow

marination committed:
Merge branch 'version-13-hotfix' into e-commerce-refactor

marination committed:
fix: Font size, empty image styles, and minor cleanup

- some frappe css font variables weren't loading, remove dependency
- fixed fallback display for missing images in views
- py code cleanup (minor)

mujeerhashmi commented:
Thanks to @revant we are now able to test this PR on this [server](https://gh-imp-2021-00092.castlecraft.co.in/).

@marination I appreciate the efforts you have put in for this to come up.

My honest review after a bit of testing. Maybe some of the things are WIP.

### **The Good**
1. The shopping cart icons wobble effect on add to cart
2. +1 for Wishlist feature
3. +1 for Grid/List View Toggle

### **The Bad (Scope of Improvements)**
1. The product listing does not show the offer. (How do we manage to show price discounts ex: MRP: Rs.<s>999</s> **600**)
2. Add to Cart for items already in the cart resets the quantity to 1. The Add to Cart maybe can increment the quantity for already added items.
3. The "+" and "-" signs are not properly aligned in the shopping cart.

### **The Ugly**
1. The cart total can't be seen until you request for quotation.
2. Variant attribute selection is still in the pop-up.

I will add more review later.

mujeerhashmi commented:
When you add item with variant to the cart and click on the cart to view. It breaks, here are the logs

```
Uncaught Server Exception
There was an error building this page

Error Code: 500

Traceback (most recent call last):
  File "/home/frappe/frappe-bench/apps/frappe/frappe/website/render.py", line 49, in render
    data = render_page_by_language(path)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/website/render.py", line 178, in render_page_by_language
    return render_page(path)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/website/render.py", line 194, in render_page
    return build(path)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/website/render.py", line 201, in build
    return build_page(path)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/website/render.py", line 224, in build_page
    html = frappe.get_template(context.template).render(context)
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/jinja2/environment.py", line 1090, in render
    self.environment.handle_exception()
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/jinja2/environment.py", line 832, in handle_exception
    reraise(*rewrite_traceback_stack(source=source))
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/jinja2/_compat.py", line 28, in reraise
    raise value.with_traceback(tb)
  File "/home/frappe/frappe-bench/apps/erpnext/erpnext/./templates/pages/cart.html", line 1, in top-level template code
    {% extends "templates/web.html" %}
  File "/home/frappe/frappe-bench/apps/frappe/frappe/./templates/web.html", line 1, in top-level template code
    {% extends base_template_path %}
  File "/home/frappe/frappe-bench/apps/frappe/frappe/./templates/base.html", line 83, in top-level template code
    {% block content %}
  File "/home/frappe/frappe-bench/apps/frappe/frappe/./templates/web.html", line 72, in block "content"
    {{ main_content() }}
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/jinja2/sandbox.py", line 462, in call
    return __context.call(__obj, *args, **kwargs)
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/jinja2/runtime.py", line 679, in _invoke
    rv = self._func(*arguments)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/./templates/web.html", line 15, in template
    {% block page_container %}
  File "/home/frappe/frappe-bench/apps/frappe/frappe/./templates/web.html", line 30, in block "page_container"
    {%- block page_content -%}{%- endblock -%}
  File "/home/frappe/frappe-bench/apps/erpnext/erpnext/./templates/pages/cart.html", line 42, in block "page_content"
    {% include "templates/includes/cart/cart_items.html" %}
  File "/home/frappe/frappe-bench/apps/erpnext/erpnext/./templates/includes/cart/cart_items.html", line 34, in top-level template code
    {{ _('Variant of') }} <a href="{{frappe.db.get_value('Item', variant_of, 'route')}}">{{ variant_of }}</a>
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/jinja2/sandbox.py", line 462, in call
    return __context.call(__obj, *args, **kwargs)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/database.py", line 406, in get_value
    order_by, cache=cache, for_update=for_update)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/database.py", line 450, in get_values
    out = self._get_values_from_table(fields, filters, doctype, as_dict, debug, order_by, update, for_update=for_update)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/database.py", line 605, in _get_values_from_table
    values, as_dict=as_dict, debug=debug, update=update)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/database/database.py", line 146, in sql
    self._cursor.execute(query, values)
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/cursors.py", line 148, in execute
    result = self._query(query)
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/cursors.py", line 310, in _query
    conn.query(q)
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/connections.py", line 548, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/connections.py", line 775, in _read_query_result
    result.read()
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/connections.py", line 1156, in read
    first_packet = self.connection._read_packet()
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/connections.py", line 725, in _read_packet
    packet.raise_for_error()
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/protocol.py", line 221, in raise_for_error
    err.raise_mysql_exception(self._data)
  File "/home/frappe/frappe-bench/env/lib/python3.7/site-packages/pymysql/err.py", line 143, in raise_mysql_exception
    raise errorclass(errno, errval)
pymysql.err.OperationalError: (1054, "Unknown column 'route' in 'field list'")
Your email address...
© First Critique
Powered by ERPNext - ERP Software for Distribution Companies
```

marination committed:
fix: Checkbox sizes and paging buttons

- Show paging buttons at all times and enable/disable buttons
- Define checkbox and radio button  sizes within erpnext
- Wishlist cards even height

marination committed:
fix: Missing Image in Item Page and Variants in Cart

- Added fallback for missing image in item page
- Explore button font
- Incorrectly fetching route from Item master, use Website Item instead

marination commented:
@mujeerhashmi Thanks a ton for testing this. Yes a bit of it is still WIP. 
There was an issue with styles (variables) which has been fixed. The traceback you shared has been fixed too. Maybe the server can be updated now.

>The product listing does not show the offer

I understand it's hard to test without documentation. But you can create an item price and add a pricing rule on it (like 10% off). Make sure this pricing rule isn't coupon code based and striked prices will be visible in the listing

>Add to Cart for items already in the cart resets the quantity to 1

Absolutely valid i'll fix that

> The "+" and "-" signs are not properly aligned in the shopping cart.

This is undergoing a design and overall cleanup as of now, will look into it

>The cart total can't be seen until you request for quotation.

This was hidden during the v13 redesign for some reason, i will fix this too

>Variant attribute selection is still in the pop-up.

Unfortunately i could not fit refactoring the popup as it led to quite a lot of changes (ui and code design) that I couldn't pursue due to constraints. I will continue to pursue this, thinking I'd take it up in the next phase. Limiting the scope for this PR. Sorry about that

ossplus commented:
Step:
1. Guest user visit the product list page and click "add to cart"
2. redirect to login page
3. login as website user (test@test.com qaz31.. )
4. login success and redirect to the product list page
5. It should show the "Cart" icon on the top right of the page now. But it doesn't show it.
![image](https://user-images.githubusercontent.com/1935044/121265213-00ba2100-c8eb-11eb-8946-4cceab358bb6.png)

rtdany10 commented:
Currently, items can be added to cart only after logging in. Can that be avoided?
Prompting login only upon buying would be better.

mujeerhashmi commented:
> I understand it's hard to test without documentation. But you can create an item price and add a pricing rule on it (like 10% off). Make sure this pricing rule isn't coupon code based and striked prices will be visible in the listing

Cool. Tested and It Works.

federicocalvo commented:
> Currently, items can be added to cart only after logging in. Can that be avoided?
> Prompting login only upon buying would be better.

I believe that I have been told on other occasions that this is one of the most requested functionalities, but I believe that I have been told that it is not included in the work around.

It is one of the most important user experience issues in an ecommerce, all ecommerce spend time to constantly improve this workflow to make it easier and easier.

Without this option the ecommerce is reduced to a B2B commerce.

ossplus commented:
When will this PR be merged in v13?

marination committed:
Merge branch 'version-13-hotfix' of https://github.com/frappe/erpnext into e-commerce-refactor

marination committed:
chore: Cleanup Query Engine and Product query API

- Resolved merge conflicts in item_group.py
- Separate api.py file for product listing backend api
- Brought back ORM in query engine, handled missing cases (website item groups, etc)
- Return results from API in a descriptive manner, helps keep sanity in JS
- On toggling views store view preference in localStorage

revant commented:
https://gh-imp-2021-00092.castlecraft.co.in is upgraded

marination committed:
fix: Filters state, Search input clearing, Paging buttons

- Fixed repetitive calls on checking filter checkbox
- Query count of items after offset for accurate Paging button display
- Order items by ranking in query
- Search results get empty on clearing input

marination committed:
feat: Guest Display Settings

- Re-arranged Settings checkboxes
- Deleted `show_availability_status` from E Commerce Settings
- Added `Hide Price for Guest` functionality
- Added Custom Redirection for Guest Actions in  E Commerce Settings
- Minor cleanups

marination committed:
Merge branch 'version-13-hotfix' of https://github.com/frappe/erpnext into e-commerce-refactor

marination committed:
fix: Sync Website Items on Item Change, Short Website Description

- Added Short Description in Website Item for List View
- Update Website Item on Item info change
- Un-publish Web Item if Item is disabled
- Removed unnecessary dependency on Item from query engine
- Rearranged item detail fields in Website Item
- Added Link to Website Item on Item Dashboard

marination committed:
fix: Rename Wishlist Item, Use Website Item Name in Wishlist Cards

- Renamed Wishlist Items to Wishlist Item
- Use Website Item Name in Wishlist
- Add Website Item Name field in Wishlist Item
- Remove accidental extra `font-size` attribute

marination committed:
fix: Discount Filtes & Filter behaviour

- Client: Maintain state where listing is re-rendered due filter trigger
- Client: Handle binding/restoring discount filters separately on filter trigger
- Client: Placeholder Image for search results
- If any filter is checked, query and display items from page 1
- Query Engine: Smaller functions and handle discount filter properly
- Added index on item group and brand for Website item

marination committed:
chore: UI refresh for grid/list view and search

- enhanced UI for grid/list view, actions visible on hover only
- enhanced search UI
- Added indicator to show if item is in cart
- Moved search with view togglers

marination committed:
feat: Recommended Items and Item full page refresh

- Added Optional Recommended Items
- Item Full Page minor UI Refresh
- Floating wishlist button in item full page
- Reviews section UI Refresh

marination committed:
chore: Wishlist UI (minor)

- Minor wishlist UI refresh, actions on hover, new icon
- Increase max length of wishlist card title
- Dont fetch outdated price in wishlist
- Translate 'out of stock'
- Use ORM

marination committed:
feat: Cart minor UI/UX Refresh

- Added Setting to show or hide price if checkout is disabled
- Show Web Item name in cart instead of Desk Item name
- Cart minor UI Refresh: added images in cart
- Cart minor UI Refresh: repositioned remove button and redesigned
- Cart minor UI Refresh: Payment Summary section
- Cart minor UI Refresh: Disable input on free item
- Cart minor UI Refresh: Add address button in cards
- New file for cart payment summary UI with coupon code (old)

marination committed:
Merge branch 'version-13-hotfix' of https://github.com/frappe/erpnext into e-commerce-refactor

marination committed:
Merge branch 'version-13-hotfix' into e-commerce-refactor

marination committed:
fix: Remove Website Item from Item Dashboard

- Install breaks on CI, will add later

marination committed:
test: Website Item basic test & removed product configurator tests

- Sider fixes: missing semicolons
- Added two basic tests for Website Item
- Commented Product Configurator tests, needs to re-written fully

alaahariri commented:
Hi All, 

You are doing a great efforts.

I have a question? When this will be merged to Production branch?

marination commented:
@alaahariri @gmplab hey this PR is awaiting review, tests are still being written, we'll mostly take a separate release for this (unsure). You can expect it before the end of next month.

alaahariri commented:
> @alaahariri @gmplab hey this PR is awaiting review, tests are still being written, we'll mostly take a separate release for this (unsure). You can expect it before the end of next month.

@marination Thank you so much for your reply. Please let me know if you ever need any testing efforts from me. I can be of a good tester.

surajshetty3416 commented on erpnext/e_commerce/product_search.js:
```diff
@@ -0,0 +1,244 @@
+		this.bindSearchAction();
+	}
+
+	setupSearchDropDown() {
```
```suggestion
	setup_search_dropdown() {
```
Why camel case in this file?

marination replied:
Was following general JS case, wasn't sure if i should keep it consistent with the case in frappe since its an independent JS file. Will change it

marination committed:
test: Website Item (Item full page)

- Tests for website item. Desk tests and portal tests
- Allow guests on method `get_offer_details`
- Fetch stock details only if `show_stock_availability` is enabled
- Validation for duplicate web items on item merge
- Separate method for `validate_properties_before_merge`
- Common error title and exception `DataValidationError` for merge validations on Item

marination committed:
Merge branch 'version-13-hotfix' of https://github.com/frappe/erpnext into e-commerce-refactor

marination committed:
test: Recommendations, Reviews and Wishlist

- Tests for verified and unverified item reviewers
- Test for recommended items and their prices
- Test for adding removing items from Wishlist
- Bug: Wishlist deletes all entries of an item code irrespective of user
- Get Item reviews only if enabled
- Removed price fields from Wishlist Item and made fields read only
- Removed unused price stored as data on Wishlist buttons
- Customer Reviews page checks if reviews are enabled else shows No Reviews
- Moved price stock fetching in Wishlist in separate function
- Made fields read only in Item Review

marination committed:
Merge branch 'version-13-hotfix' into e-commerce-refactor

marination committed:
chore: debug travis

marination committed:
chore: Re-organise files,remove T&C modal in cart

- Moved product query and filters engine to `product_data_engine` folder
- Moved product grid, list, search, view to `product_ui` folder
- Renamed `website_item_indexing.py` to `redisearch.py`
- Render Terms and Conditions server side along with the rest of the Shopping cart. Don’t make another db call
- Style changes to terms and conditions
- Deleted unused `cart_terms.html`
- Removed print statements

marination committed:
Merge branch 'version-13-hotfix' into e-commerce-refactor

marination committed:
fix: Error state and passing args for product listing

- Show error state in case of unexpected errors in query engine
- Pass args appropriately from `view.js`
- Use args correctly in `api.py`
- Error handling in `api.py` if query engine raises error
- Instantiated product data engine tests
- Fix dotted path for search api call in `search.js`

marination committed:
test: Product Query & Filter Engine, Item Group Page

- Test for ProductQuery engine and ProductFilters engine
- Test for engine for Item Group too
- Renamed ‘product_configurator’ to ‘variant_selector’
- Cleaned up filters.py
- Modal freeze backdrop lighter only in cart, since there’s nothing over it
- Fixed unusual spacing in variant selector dialog
- Made `get_child_groups_for_website` more readable
- Replaced ‘Configure’ with ‘Select’ for variant selection

surajshetty3416 commented on erpnext/e_commerce/doctype/e_commerce_settings/e_commerce_settings.py:
```diff
@@ -1,25 +1,85 @@
+class ECommerceSettings(Document):
 	def onload(self):
 		self.get("__onload").quotation_series = frappe.get_meta("Quotation").get_options("naming_series")
+		self.is_redisearch_loaded = is_search_module_loaded()
```
where is this used?

surajshetty3416 commented on erpnext/e_commerce/doctype/e_commerce_settings/e_commerce_settings.py:
```diff
@@ -93,3 +164,9 @@ def check_shopping_cart_enabled():
 def show_attachments():
 	return get_shopping_cart_settings().show_attachments
+
+def home_page_is_products(doc, method):
```
```suggestion
def set_home_page(doc, method):
```

surajshetty3416 commented on erpnext/e_commerce/doctype/item_review/item_review.py:
```diff
@@ -0,0 +1,92 @@
+			reviews_per_rating.append(percent)
+
+		data.reviews_per_rating = reviews_per_rating
+		data.total_reviews = rating_data.total
```
Is this function called every time we load a product page? if yes, then let's cache review rating stats. We can burst cache on new ratings.

surajshetty3416 commented on erpnext/e_commerce/doctype/e_commerce_settings/e_commerce_settings.py:
```diff
@@ -93,3 +164,9 @@ def check_shopping_cart_enabled():
 def show_attachments():
 	return get_shopping_cart_settings().show_attachments
+
+def home_page_is_products(doc, method):
```
```suggestion
def set_home_page(doc, method):
```
Also, can we add an alert to inform the user about this internal magic to avoid homepage reset frustration?

marination replied:
@surajshetty3416 I'll remove this checkbox and this reset feature. Users can simply set it in Website Settings. All it does is set the home page to `products` thats it

surajshetty3416 commented on erpnext/e_commerce/doctype/e_commerce_settings/e_commerce_settings.json:
```diff
@@ -0,0 +1,415 @@
+   "fieldname": "redisearch_warning",
+   "fieldtype": "HTML",
+   "label": "Redisearch Warning",
+   "options": "<p class=\"alert alert-warning\">Redisearch module not loaded. If you want to use advanced product search features, refer documentation <a class=\"alert-link\" href=\"https://frappeframework.com/docs/user/en/installation\" target=\"_blank\">here</a>.</p>"
```
do we have Redis search module installation docs?

marination replied:
@surajshetty3416 https://github.com/frappe/erpnext_documentation/pull/375, will change the link, got shifted to erpnext

surajshetty3416 commented on erpnext/e_commerce/doctype/item_review/item_review.py:
```diff
@@ -0,0 +1,92 @@
+def add_item_review(web_item, title, rating, comment=None):
+	""" Add an Item Review by a user if non-existent. """
+	if frappe.session.user == "Guest":
+		frappe.throw(_("You are not verified to write a review yet. Please contact us for verification."),
```
how to contact? Maybe we can link the "contact us" page?

marination replied:
this should not be thrown ever ideally, unless someone calls this via an API as a Guest. The button that triggers this is hidden if user is a Guest, so this is defensive.

surajshetty3416 commented on erpnext/e_commerce/doctype/website_item/website_item.py:
```diff
@@ -0,0 +1,504 @@
+			context.attribute_values = {}
+			context.selected_attributes = {}
+
+			# load attributes
+			for v in context.variants:
+				v.attributes = frappe.get_all("Item Variant Attribute",
+					fields=["attribute", "attribute_value"],
+					filters={"parent": v.name})
+				# make a map for easier access in templates
+				v.attribute_map = frappe._dict({})
+				for attr in v.attributes:
+					v.attribute_map[attr.attribute] = attr.attribute_value
+
+				for attr in v.attributes:
+					values = attribute_values_available.setdefault(attr.attribute, [])
+					if attr.attribute_value not in values:
+						values.append(attr.attribute_value)
+
+					if v.name == context.variant.name:
+						context.selected_attributes[attr.attribute] = attr.attribute_value
+
+			# filter attributes, order based on attribute table
+			item = frappe.get_cached_doc("Item", self.item_code)
+			for attr in item.attributes:
+				values = context.attribute_values.setdefault(attr.attribute, [])
+
+				if cint(frappe.db.get_value("Item Attribute", attr.attribute, "numeric_values")):
+					for val in sorted(attribute_values_available.get(attr.attribute, []), key=flt):
+						values.append(val)
+
+				else:
+					# get list of values defined (for sequence)
+					for attr_value in frappe.db.get_all("Item Attribute Value",
+						fields=["attribute_value"],
+						filters={"parent": attr.attribute}, order_by="idx asc"):
+
+						if attr_value.attribute_value in attribute_values_available.get(attr.attribute, []):
+							values.append(attr_value.attribute_value)
+
+			context.variant_info = json.dumps(context.variants)
```
please try and break such big functions into smaller ones.

surajshetty3416 commented on erpnext/e_commerce/doctype/wishlist/wishlist.py:
```diff
@@ -0,0 +1,65 @@
+@frappe.whitelist()
+def remove_from_wishlist(item_code):
+	if frappe.db.exists("Wishlist Item", {"item_code": item_code, "parent": frappe.session.user}):
+		frappe.db.sql("""
+			DELETE
+			FROM `tabWishlist Item`
+			WHERE
+				item_code=%(item_code)s
+				and parent='%(user)s'
+		""" % {"item_code": frappe.db.escape(item_code), "user": frappe.session.user})
```
why not `frappe.db.delete`?

marination committed:
Merge branch 'version-13-hotfix' of https://github.com/frappe/erpnext into e-commerce-refactor

marination committed:
chore: Remove `Home Page is Products` and Sider fixes

- Removed `Home Page is Products` checkbox in E Comm Settings. Can be manually set in Website Settings
- Removed hooks trigger to reset home page as products
- Sider: duplicate color attribute, shift `return` to next line, over-indentation

marination committed:
refactor: Cache Item Reviews and other review feedback

- `get_doc` -> `get_values` and `db.sql` -> `db.delete` in Wishlist Item deletion
- cache first page of Item Reviews and burst cache on addition and deletion of reviews
- Update redisearch docs link in E Commerce Settings
- Removed unused cint import
- Broke setting attribute context into smaller functions and code cleanup
- Minor recommended items padding tweak
- Item reviews form dict now uses website item as key
- Customer reviews rendered from UI style consistency
- Stock status consistency in listing and full page
- Handle no price in variant dialog for matched item

marination committed:
Merge branch 'version-13-hotfix' into e-commerce-refactor

marination committed:
Merge branch 'version-13-hotfix' of https://github.com/frappe/erpnext into e-commerce-refactor

marination committed:
fix: Sider and CI

- Incorrect path and Settings in price_list.py import due to merge
- Extra space removed in website_item.py
- Changed client side namespace to `erpnext.e_commerce.*`

marination committed:
Merge branch 'version-13-hotfix' of https://github.com/frappe/erpnext into e-commerce-refactor

marination committed:
Chore: Miscellaneous UI review changes

- Added bg (variable) to pages, card space separation visible
- Removed `show brand line` in settings, shown by default
- Re-arranged settings by importance
- View toggling primary colour is grey
- Only populate recent searches on successful search
- Hit only one server side api, once while searching
- List view primary button float right
- Discounts takes upper limit eg. 10% and below
- Navbar icons only wiggle on qty increase in cart/wishlist
- Pay button in SO portal
- Remove bottom white space below item full page image, min-height fits to content
- List view vertical space between heading and item code
- Empty offer subtitle handled

marination committed:
Merge branch 'version-13-hotfix' into e-commerce-refactor

marination committed:
feat: (minor) Backorder indicator and fixed inconsistencies

- Checkbox in website item to indicate if item is on backorder
- Indicator on listing on full page if availbale on backorder.
- fix: Allow provision to add any valid field from Website Item in Search Index
- fix: Settings filter fields are as per Item, make as per Website Item
- "Add to quote/ Go to Quote" if cart checkout is disabled

marination committed:
fix: Failing search util import and patch

- Use simple function import instead of global variable to check if field is valid to index
- Commit after every 20 items in patch to  create web items from items

marination committed:
test: Set Price list and company in data engine test setup

marination committed:
Merge branch 'version-13-hotfix' into e-commerce-refactor

ankush commented:
@mergifyio backport develop
